### PR TITLE
Work In Progress: add speaker model

### PR DESCRIPTION
**Pull request by djds23**
Progress on #100

Comments welcome!
### TODOs
- [x] Hammer out what exactly is a primary speaker, and how do we represent it
- [x] properly validate emails being saved in speaker invite
- [ ] send emails to speaker invited emails
- [x] change claimable token from random UUID, to some sort of signed token
- [x] fix relationships between speakers, talks and users (backref -> secondary) 
- [ ] use fieldlist to actually add an array of fields, that way we can truly have N co-presenters, rather than one extra.
- [ ] fix the optional validation on co-presenter email
- [ ] add claimable endpoint/template
#### Pipe dream
- [ ] tests!
- [ ] Admin pages for resending emails to speaker invites

### CODE REVIEW

File: pygotham/frontend/talks.py [hunk 1]
```diff
@@ -100,6 +100,13 @@ def proposal(pk=None):
         db.session.add(talk)
         db.session.commit()
 
+        speaker = Speaker(
```
Comment (dirn) [inline pygotham/frontend/talks.py]: Do we want to add the primary speaker to the speakers table? If so, we probably want to remove some fields from the talks table (which will require a data migration). We'll also need some way of tracking the primary speaker.

File: pygotham/talks/models.py [hunk 2]
```diff
@@ -1,11 +1,12 @@
 """Talks models."""
 
 from slugify import slugify
+from sqlalchemy_utils import ArrowType
 
 from pygotham.core import db
 from pygotham.events.query import EventQuery
 
-__all__ = ('Category', 'Duration', 'Talk')
+__all__ = ('Category', 'Duration', 'Talk', 'Speaker')
```
Comment (dirn) [inline pygotham/talks/models.py]: `S` comes before `T`.

File: pygotham/frontend/talks.py [hunk 3]
```diff
@@ -100,6 +100,13 @@ def proposal(pk=None):
...
+            user_id=current_user.id,
+            talk_id=talk.id,
+        )
+        db.session.add(speaker)
+        db.session.commit()
```
Comment (jonafato) [inline pygotham/frontend/talks.py]: We should create the speaker and add it to the session before calling ``db.session.commit()`` above.

1. This will make sure that we're not making extra commits when not necessary.
2. Additionally, it will prevent talks from being submitted without users (here, `Speaker` instances) assigned to them; if something fails causing that second commit to error, we'll want to make sure that we don't create an "orphaned" talk in the database with no user to take credit for it.

Finally, when removing the commit that happens between speaker and talk creation, `talk_id=talk.id` should change to `talk=talk`. SQLAlchemy is smart enough to track commit dependency and assign the correct id on the fly, so we don't need to keep track of ids and can work with high level Python objects.

File: pygotham/talks/models.py [hunk 4]
```diff
@@ -107,3 +108,30 @@ def is_accepted(self):
...
+    )
+    talk = db.relationship(
+        'Talk', backref=db.backref('talks', lazy='dynamic'),
+    )
+
+    user_id = db.Column(
+        db.Integer, db.ForeignKey('users.id'), nullable=False
+    )
+    user = db.relationship(
+        'User', backref=db.backref('users', lazy='dynamic')
```
Comment (dirn) [inline pygotham/talks/models.py]: This backref is a bit confusing. What are you expecting `user.users` to provide? We're probably better off having a back ref on either `Talk` or `User` that uses `secondary` to go all the way to `Talk` (i.e., `user.talks` and `talk.speakers`).

File: pygotham/frontend/talks.py (hunk 1 above)
Comment (djds23) [inline pygotham/frontend/talks.py]: I am not sure we should have a concept of the 'primary speaker'. What would that afford us? Also, what beyond 'user_id' & 'recording_release' would we want to drop from the talk model?

File: pygotham/talks/models.py (hunk 4 above)
Comment (jonafato) [inline pygotham/talks/models.py]: `secondary` is probably the correct behavior here. I think we'll most often reference things like `user.talks` and `talk.speakers`, so this makes things significantly less verbose.

File: pygotham/frontend/talks.py (hunk 1 above)
Comment (dirn) [inline pygotham/frontend/talks.py]: Are we planning to let all co-presenters edit the talk? We'll need to figure out how we handle two people editing the same talk at the same time.

Comment (djds23) [inline pygotham/frontend/talks.py]: That is a fantastic point, do we know how pygotham.org /  other conferences handle this? I think we should let all of the speakers edit talk information. If you invite someone to give a talk with you, then you likely trust them to modify the content of said talk.

Comment (dirn) [inline pygotham/frontend/talks.py]: pygotham.org currently doesn't handle this :wink:

Comment (jonafato) [inline pygotham/frontend/talks.py]: A primary speaker ensures that someone that I invite as a copresenter can't come along and steal my talk away from me. A boolean field on the speaker table probably satisfies this issue (along with some check constraint that makes sure we don't end up with more than one primary speaker for a given talk).

`recording_release` should probably become a property that computes the `AND` of all of a talk's speakers' recording releases rather than be removed entirely.

File: pygotham/frontend/talks.py (hunk 3 above)
Comment (djds23) [inline pygotham/frontend/talks.py]: God bless ORMs.

File: pygotham/frontend/talks.py [hunk 5]
```diff
@@ -93,18 +94,29 @@ def proposal(pk=None):
...
+            user=current_user,
+            talk=talk,
+        )
         db.session.add(talk)
+        db.session.add(speaker)
+
+        for invite_form in talk_form.speaker_invites:
+            if invite_form.validate(invite_form):
+                speaker_invite = SpeakerInvite(talk=talk, **invite_form.data)
+                db.session.add(speaker_invite)
```
Comment (djds23) [inline pygotham/frontend/talks.py]: @jonafato & @dirn I am convinced this is not the right way to do this, but I can't find a better way. `populate_obj` wasn't working. Also line 109 seems strange, if there is another way todo it please let me know.

Comment (dirn) [inline pygotham/frontend/talks.py]: What does "wasn't working" mean? Were you getting an exception?

Comment (djds23) [inline pygotham/frontend/talks.py]: great question. I was getting an exception because `populate_obj` on the `invite_form` appears to require some sort of `name` positional arg after the object.

File: migrations/versions/4499b94e5e6_added_speakers_table.py [hunk 6]
```diff
@@ -0,0 +1,38 @@
+"""Added speakers table
```
Comment (dirn) [inline migrations/versions/4499b94e5e6_added_speakers_table.py]: Small nit: migration descriptions should be like commit subjects; they should describe what they do as they run. The tense should be changed here to "add" rather than "added." Since this migration will need to be updated to include a data migration as well, mentioning each aspect of the migration in the description can make it overly chatting. Focusing on the bigger picture helps.

> Add co-presenters.

The period was included to be consistent with other migrations (some are missing it, but most have it).

Comment (djds23) [inline migrations/versions/4499b94e5e6_added_speakers_table.py]: Noted! Thanks for the comment, I will update it with these comments in mind

File: migrations/versions/492839e0f6a_added_speaker_invites_table.py [hunk 7]
```diff
@@ -0,0 +1,33 @@
+"""Added speaker_invites table
```
Comment (dirn) [inline migrations/versions/492839e0f6a_added_speaker_invites_table.py]: This migration should be squashed into the first one. If doing that now messes up your progress, it's okay to hold off on squashing them until the end, but you probably don't want too many of them to accumulate before you do that.

File: pygotham/frontend/talks.py [hunk 8]
```diff
@@ -9,7 +9,8 @@
 
 from pygotham.core import db
 from pygotham.frontend import direct_to_template, route
-from pygotham.models import Day, Talk
+from pygotham.models import Day, Talk, Speaker
```
Comment (dirn) [inline pygotham/frontend/talks.py]: Another nit: alphabetize your imports, `Speaker` before `Talk`.

File: migrations/versions/492839e0f6a_added_speaker_invites_table.py (hunk 7 above)
Comment (djds23) [inline migrations/versions/492839e0f6a_added_speaker_invites_table.py]: Good call! I will start squashing before I add any more changes.

File: pygotham/frontend/templates/talks/proposal.html [hunk 9]
```diff
@@ -19,6 +19,10 @@
         {{ wtf.horizontal_field(form.duration, placeholder='Duration') }}
         {{ wtf.horizontal_field(form.abstract, placeholder='Abstract') }}
         {{ wtf.horizontal_field(form.outline, placeholder='Outline') }}
+        {% for invite_form in form.speaker_invites %}
+          {{ invite_form.hidden_tag() }}
+          {{ wtf.horizontal_field(invite_form.invited_email, placeholder='your.partner@in.crime') }}
```
Comment (dirn) [inline pygotham/frontend/templates/talks/proposal.html]: The placeholder should be consistent with the others, something like `Email` or `Co-presenter email`.

Comment (djds23) [inline pygotham/frontend/templates/talks/proposal.html]: I was trying to get cutesy with the copy :P, will change!

File: pygotham/talks/forms.py [hunk 10]
```diff
@@ -13,10 +13,23 @@
 
 
 def duration_query_factory():
-    """Return available :class:`~pygotha.models.Duration` instances."""
+    """Return available :class:`~pygotham.models.Duration` instances."""
```
Comment (dirn) [inline pygotham/talks/forms.py]: As silly as this sounds, try to avoid making unrelated changes under a pull request. If we had this feature and at some point decide to revert it, this typo will return.

File: pygotham/talks/forms.py [hunk 11]
```diff
@@ -13,10 +13,23 @@
...
     return Duration.query.filter(Duration.inactive == False)
 
 
+class SpeakerInvitesForm(ModelForm):
+
```
Comment (dirn) [inline pygotham/talks/forms.py]: The docstring goes here.

File: pygotham/talks/forms.py (hunk 10 above)
Comment (djds23) [inline pygotham/talks/forms.py]: very true! I will do this separately

File: pygotham/talks/models.py [hunk 12]
```diff
@@ -43,6 +46,56 @@ def __str__(self):
...
+    talk = db.relationship(
+        'Talk', backref=db.backref('speakers', lazy='dynamic'),)
+
+    user_id = db.Column(
+        db.Integer, db.ForeignKey('users.id'), nullable=False)
+    user = db.relationship(
+        'User', backref=db.backref('speakers', lazy='dynamic'))
+
+    recording_release = db.Column(db.Boolean, nullable=True)
+    confirmed_ts = db.Column(ArrowType, nullable=True)
```
Comment (dirn) [inline pygotham/talks/models.py]: Don't include data types in column names. Just as you wouldn't call a column `id_int` you shouldn't include the `_ts` prefix. You should also look for prior art when naming things. `confirmed_at` is used elsewhere. You'd then use `declined_at` for the other field.

File: pygotham/users/models.py [hunk 13]
```diff
@@ -20,6 +20,8 @@
     db.Column('role_id', db.Integer, db.ForeignKey('roles.id')),
 )
 
+user_talks = db.Table('speakers', Speaker.metadata, autoload=True)
```
Comment (dirn) [inline pygotham/users/models.py]: Moar nits: Table names are pluralized, so their corresponding object names should match: `users_talks`. (I realize the table name here is different, but it's serving the same purpose here.)

Comment (dirn) [inline pygotham/users/models.py]: Also, is this table object even necessary? Can't `Speaker` be used as the value of `secondary` in the relationship between `User` and `Talk`?

File: pygotham/frontend/templates/talks/proposal.html (hunk 9 above)
Comment (dirn) [inline pygotham/frontend/templates/talks/proposal.html]: I'm generally in favor of cutesy, but we don't do that anywhere else. In my head, consistency is better.

**Comment by jonafato** [comment]
A couple of questions:
> change claimable token from random UUID, to some sort of signed token

What will be signed here? If we're going to store something, a random token should suffice. One of the benefits of using a signature is that we can omit storing a token entirely. I'm tempted to use something like an [`itsdangerous.URLSafeTimedSerializer`](https://pythonhosted.org/itsdangerous/#itsdangerous.URLSafeTimedSerializer) here to sign and serialize the combination of the invited email address and the invite ID; this avoids managing a database full of tokens entirely.

> Admin pages for resending emails to speaker invites

What "admin" pages would be included? Would this be actually part of PyGotham's Flask-Admin interface? If so, it's probably a lot of work for not much benefit. It would be nice, however, to have a button on the talk edit page to allow primary speakers to resend invites.

File: pygotham/talks/forms.py [hunk 14]
```diff
@@ -13,10 +13,23 @@
...
+    class Meta:
+        model = SpeakerInvite
+        exclude = ('claim_token',)
+        field_args = {
+            'invited_email': {
+                'label': 'Co-Presenter\'s Email',
```
Comment (jonafato) [inline pygotham/talks/forms.py]: Nit: avoid escaping quotes when possible (`"Co-Presenter's Email"`).

File: pygotham/talks/models.py [hunk 15]
```diff
@@ -90,7 +143,10 @@ class Talk(db.Model):
     )
 
     user_id = db.Column(db.Integer, db.ForeignKey('users.id'), nullable=False)
-    user = db.relationship('User', backref=db.backref('talks', lazy='dynamic'))
+    speaking_users = db.relationship(
```
Comment (jonafato) [inline pygotham/talks/models.py]: I think we can just call this `users`.

File: pygotham/talks/models.py [hunk 16]
```diff
@@ -90,7 +143,10 @@ class Talk(db.Model):
     )
 
     user_id = db.Column(db.Integer, db.ForeignKey('users.id'), nullable=False)
```
Comment (jonafato) [inline pygotham/talks/models.py]: When we move everything to the `Speaker` model, we should remove `user_id` from `Talk`.

Comment (dirn) [inline pygotham/talks/models.py]: That same applies to `recording_release`, although that should be replaced by a property.

Comment (jonafato) [inline pygotham/talks/models.py]: @dirn On that note, it may be useful to have a `primary_user` property, e.g. to determine if a user is able to edit a talk.

Comment (djds23) [inline pygotham/talks/models.py]: I think it might be useful to just rename the `user` column to `primary_user`.

Comment (dirn) [inline pygotham/talks/models.py]: Is this in addition to or instead of storing it in the speakers table?

File: pygotham/users/models.py (hunk 13 above)
Comment (djds23) [inline pygotham/users/models.py]: I tried, and I get this exception:

```python
  File "/Users/deansilfen/.virtualenvs/pygotham/lib/python3.4/site-packages/sqlalchemy/sql/selectable.py", line 707, in _join_condition
    b.foreign_keys,
AttributeError: 'Speaker' object has no attribute 'foreign_keys'
```

File: pygotham/talks/models.py (hunk 15 above)
Comment (djds23) [inline pygotham/talks/models.py]: I agree

File: pygotham/users/models.py (hunk 13 above)
Comment (jonafato) [inline pygotham/users/models.py]: Per the docs (the code sample immediately before [this section](http://docs.sqlalchemy.org/en/latest/orm/basic_relationships.html#deleting-rows-from-the-many-to-many-table)), you can just use the table name here.

Comment (djds23) [inline pygotham/users/models.py]: done!

File: pygotham/frontend/talks.py [hunk 17]
```diff
@@ -115,6 +133,11 @@ def proposal(pk=None):
 )
 
 
+@route(blueprint, '/confirm/',  methods=('GET', 'POST'))
+def confirm():
+    current_user.email, current_user.id
```
Comment (djds23) [inline pygotham/frontend/talks.py]: This is not finished! carry on

File: pygotham/frontend/talks.py [hunk 18]
```diff
@@ -94,10 +98,24 @@ def proposal(pk=None):
...
     form = TalkSubmissionForm(obj=talk)
+
     if form.validate_on_submit():
         form.populate_obj(talk)
 
+        speaker = Speaker(
+            user=current_user,
+            talk=talk,
+            primary=True,
+            confirmed_at=datetime.now(),
```
Comment (jonafato) [inline pygotham/frontend/talks.py]: Please use `arrow.utcnow().to('America/New_York').naive` to be consistent with other parts of the codebase here.

Comment (djds23) [inline pygotham/frontend/talks.py]: good catch

File: pygotham/frontend/profile.py [hunk 19]
```diff
@@ -20,7 +20,19 @@
...
-    talks = Talk.query.current.filter(Talk.user == current_user)
+    talks = db.session.query(
+        Talk,
+        Speaker,
+    ).join(
+        'speakers'
+    ).filter(
+        Speaker.user == current_user
+    ).filter(
+        Speaker.declined_at is None
```
Comment (llllllllll) [inline pygotham/frontend/profile.py]: I think this wants to be `.is_(None)` or `== None`, `is` cannot be overloaded so sqla uses a method or `==` to check for `NULL`

**Comment by dirn** [comment]
@djds23, I'm going to close this pull request. If you'd like to pick it back up, feel free to do so, but since it's been over a year since any work has been done, I'd like to give others a chance to work on it.
